pkg/listwatch/listwatch.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2016 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package listwatch
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/cache"
+)
+
+// NewUnprivilegedNamespaceListWatchFromClient mimics cache.NewListWatchFromClient.
+// It allows for the creation of a cache.ListWatch for namespaces from a client that does not have `List` privileges.
+// If the slice of namespaces contains only v1.NamespaceAll, then this func assumes that the client
+// has List and Watch privileges and returns a regular cache.ListWatch, since there is no other way to get all namespaces.
+func NewUnprivilegedNamespaceListWatchFromClient(c cache.Getter, namespaces []string, fieldSelector fields.Selector) *cache.ListWatch {
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = fieldSelector.String()
+	}
+	return NewFilteredUnprivilegedNamespaceListWatchFromClient(c, namespaces, optionsModifier)
+}
+
+// NewFilteredUnprivilegedNamespaceListWatchFromClient mimics cache.NewUnprivilegedNamespaceListWatchFromClient.
+// It allows for the creation of a cache.ListWatch for namespaces from a client that does not have `List` privileges.
+// If the slice of namespaces contains only v1.NamespaceAll, then this func assumes that the client
+// has List and Watch privileges and returns a regular cache.ListWatch, since there is no other way to get all namespaces.
+func NewFilteredUnprivilegedNamespaceListWatchFromClient(c cache.Getter, namespaces []string, optionsModifier func(options *metav1.ListOptions)) *cache.ListWatch {
```

Comment:
Does this function need to be exported?

Reply:
No, strictly speaking it does not. I just wanted to keep this as consistent as possible with the mentioned functions in https://github.com/kubernetes/client-go/blob/master/tools/cache/listwatch.go. There are two exported funcs, `New*` and `NewFiltered*` and the former wraps the latter. If you feel strongly about it then I am happy to change it